chips/e310x/src/chip.rs

```diff
@@ -163,7 +169,25 @@ unsafe fn handle_interrupt(intr: mcause::Interrupt) {
             CSR.mie.modify(mie::mtimer::CLEAR);
         }
         mcause::Interrupt::MachineExternal => {
+            // We recieved an interrupt, disable interrupts while we handle them
             CSR.mie.modify(mie::mext::CLEAR);
+
+            // Claim the interrupt, unwrap() as we know an interrupt exists
+            // Once claimed this interrupt won't fire until it's completed
+            // NOTE: The interrupt is no longer pending in the PLIC
+            loop {
+                let interrupt = PLIC.next_pending();
+
+                match interrupt {
+                    Some(irq) => {
+                        PLIC.save_interrupt(irq);
```

Comment:
I think this use of the global PLIC struct is unsound. Even with the guarantee that the main thread will not modify the array in a critical section. We call `save_interrupt()` from one thread (interrupt context) then read the saved state from another thread (non-interrupt context). `Cell` is not sufficient to ensure correct results will be read in the non-interrupt context after an interrupt fires.

Cell is not intended as a concurrency primitive. For the following code snippet

```rust
let cell: Cell<u32> = Cell::new(0);

let b = cell.get();
// various stuff that does not touch `cell`
let c = cell.get();
```
the compiler is allowed to assume that cell has not changed between the two loads, and elide the load that would happen when assigning `c` and instead use the result from the load when `b` was assigned. If an interrupt arrives in the interim and modified the cell, the value assigned to `c` would be incorrect.

In the case of saved interrupts, I think this could lead to extremely nefarious bugs.

This does not even consider the possibility of preemption of interrupt handlers -- which may not be possible in Tock on the HiFive? I am not sure? -- but if possible would lead to additional opportunities for bugs here as `save_interrupt` is not called atomically.

Reply:
We only ever read and write the `saved` `Cell` with interrupts disabled (I just fixed a bug where interrupts could be enabled too early), shouldn't that be enough to ensure it's not unsound?

Reply:
That protects against the potential for memory errors from the `Cell` being modified partway through a nonatomic read/write. But I don't think it protects against the compiler reordering or eliding loads, because AFAIK enabling/disabling interrupts is not a memory fence. If interrupts are disabled during all accesses to the `saved` `Cell`, I think using `VolatileCell` would be sufficient. Alternatively, using an atomic type would work, but that has other issues with forcing us to stay on nightly for `earlgrey` (which may be a problem we want to postpone to later, idk).

Reply:
After some additional thought: If you never read or write the `saved` `Cell` with interrupts enabled, this may be okay. But I think there are several calls right now that read the cell without disabling interrupts first. 

Also, I generally don't like that future modifications to this code would have to be aware that the PLIC struct is shared beyond the kernel thread into interrupt context - I don't think we currently share Rust structs between the kernel thread and interrupt context anywhere else in Tock, and doing so adds all sorts of implicit constraints on how you can interact with those objects safely (e.g. having to disable interrupts when accessing fields that can change) which might easily be missed in future modifications to this code.

Reply:
Ok, I have pushed an update that uses `VolatileCell` (good call on that one). I have changed some of the PLIC functions to be `unsafe` so that no one calls them with interrupts still enabled.